components/service/glean/src/main/java/mozilla/components/service/glean/net/HttpPingUploader.kt

```diff
@@ -62,46 +62,56 @@ internal class HttpPingUploader : PingUploader {
      *         or faced an unrecoverable error), false if there was a recoverable
      *         error callers can deal with.
      */
-    @Suppress("ReturnCount", "MagicNumber")
     override fun upload(path: String, data: String, config: Configuration): Boolean {
         logPing(path, data, config)
 
-        var connection: HttpURLConnection? = null
-        try {
-            connection = openConnection(config.serverEndpoint, path)
-            connection.requestMethod = "POST"
-            connection.connectTimeout = config.connectionTimeout
-            connection.readTimeout = config.readTimeout
-            connection.doOutput = true
+        val request = buildRequest(path, data, config)
 
-            connection.setRequestProperty("Content-Type", "application/json; charset=utf-8")
-            connection.setRequestProperty("User-Agent", config.userAgent)
-            connection.setRequestProperty("Date", createDateHeaderValue())
+        return try {
+            performUpload(config.client, request)
+        } catch (e: IOException) {
+            logger.warn("IOException while uploading ping", e)
+            false
+        }
+    }
 
+    @VisibleForTesting(otherwise = PRIVATE)
+    internal fun buildRequest(path: String, data: String, config: Configuration) = Request(
```

Comment:
Oh, wow, I didn't know we could use both `internal` and `VisibleFortesting` one one function. Has this any advantage over using only one or the other?

Reply:
I'm not sure but I think that the 'internal' would override the annotation, from my understanding of how the @VisibleForTesting annotation works.  The `otherwise` should probably be `INTERNAL` or omitted, since it defaults to `PRIVATE`.

Reply:
`internal` makes this accessible from everywhere in this gradle module. Marking it as `VisibleForTesting(otherwise = PRIVATE)` let's the IDE know that it should treat this (almost) as private. So if I access this somewhere (outside of tests) where this is not considered "private" the IDE will show a warning that I may be doing something wrong:

![screenshot 2019-03-06 at 18 28 41](https://user-images.githubusercontent.com/89638/53900819-b11a8a00-403d-11e9-86e9-04173f4a6cfa.png)

Reply:
Ahh, so the annotation just gets enforced by the IDE and not the compiler...

Reply:
Great! I'll file a bug for having that in our testing API as well :)